skimage/filters/_gabor.py

```diff
@@ -1,48 +1,338 @@
+import collections as coll
 import numpy as np
 from scipy import ndimage as ndi
 from .._shared.utils import assert_nD
+from warnings import warn
 
 
 __all__ = ['gabor_kernel', 'gabor']
 
 
+def _normalize(x):
+    """Normalizes an array.
+
+    Parameters
+    ----------
+    x : array_like
+        Array to normalize.
+
+    Returns
+    -------
+    u : array
+        Unitary array.
+
+    Examples
+    --------
+    >>> x = np.arange(5)
+    >>> uX = _normalize(x)
+    >>> np.isclose(np.linalg.norm(uX), 1)
+    True
+    """
+    u = np.asarray(x)
+
+    norm = np.linalg.norm(u)
+
+    if not np.isclose(norm, 1):
+        u = u / norm
+
+    return u
+
+
+def _compute_projection_matrix(axis, indices=None):
+    """Generates a matrix that projects an axis onto the 0th coordinate axis.
+
+    Parameters
+    ----------
+    axis : (N, ) array
+        Unit vector.
+    indices : sequence of int, optional
+        Indices of the components of `axis` that should be transformed.
+        If `None`, defaults to all of the indices of `axis`.
+
+    Returns
+    -------
+    R : (N, N) array
+        Orthogonal projection matrix.
+
+    References
+    ----------
+    .. [1] Ognyan Ivanov Zhelezov. One Modification which Increases Performance
+           of N-Dimensional Rotation Matrix Generation Algorithm. International
+           Journal of Chemistry, Mathematics, and Physics, Vol. 2 No. 2, 2018:
+           pp. 13-18. https://dx.doi.org/10.22161/ijcmp.2.2.1
+    """
+    ndim = len(axis)
+
+    if indices is None:
+        indices = range(ndim)
+
+    x = axis
+    w = indices
+
+    R = np.eye(ndim)  # Initial rotation matrix = Identity matrix
+
+    # Loop to create matrices of stages
+    for step in np.round(2 ** np.arange(np.log2(ndim))).astype(int):
+        A = np.eye(ndim)
+
+        for n in range(0, ndim - step, step * 2):
+            if n + step >= len(w):
+                break
+
+            r2 = x[w[n]] * x[w[n]] + x[w[n + step]] * x[w[n + step]]
+            if r2 > 0:
+                r = np.sqrt(r2)
+                pcos = x[w[n]] / r  # Calculation of coefficients
```

Comment:
Again, `x[w[n]]` and `x[w[n + step]]`.  Since these indices are re-used, should we name them?